node/src/reactor/participating.rs

```diff
@@ -539,55 +540,181 @@ impl reactor::Reactor for Reactor {
+            latest_block_header
+        } else {
+            let initial_pre_state;
+            let finalized_block;
+
+            let chainspec = chainspec_loader.chainspec();
+            match chainspec_loader
+                .chainspec()
+                .protocol_config
+                .activation_point
+            {
+                ActivationPoint::Genesis(genesis_timestamp) => {
+                    // Check that no blocks exist in storage so we don't try to run genesis again on
+                    // an existing blockchain
+                    if let Some(first_block_header) = storage.read_block_header_by_height(1)? {
+                        return Err(Error::CannotRunGenesisOnPreExistingBlockchain {
+                            first_block_header: Box::new(first_block_header),
+                        });
+                    }
+                    let GenesisSuccess {
+                        post_state_hash,
+                        execution_effect,
+                    } = contract_runtime.commit_genesis(chainspec)?;
+                    info!("genesis chainspec name {}", chainspec.network_config.name);
+                    info!("genesis state root hash {}", post_state_hash);
+                    trace!(%post_state_hash, ?execution_effect);
+                    initial_pre_state = ExecutionPreState::new(
+                        post_state_hash.into(),
+                        0,
+                        Default::default(),
+                        Default::default(),
+                    );
+                    finalized_block = FinalizedBlock::new(
+                        BlockPayload::default(),
+                        Some(EraReport::default()),
+                        genesis_timestamp,
+                        EraId::from(0u64),
+                        0,
+                        PublicKey::System,
+                    );
+                }
+                ActivationPoint::EraId(upgrade_era_id) => {
+                    let upgrade_block_header = storage
+                        .read_switch_block_header_by_era_id(upgrade_era_id.saturating_sub(1))?
+                        .ok_or(Error::NoSuchSwitchBlockHeaderForUpgradeEra { upgrade_era_id })?;
+                    // If it's not an emergency upgrade and there is a block higher than ours, bail
+                    // because we will overwrite an existing blockchain.
+                    if chainspec.protocol_config.last_emergency_restart
+                        != Some(upgrade_block_header.era_id())
+                    {
+                        if let Some(preexisting_block_header) = storage
+                            .read_block_header_by_height(upgrade_block_header.height() + 1)?
+                        {
+                            return Err(Error::NonEmergencyUpgradeWillClobberExistingBlockChain {
+                                preexisting_block_header: Box::new(preexisting_block_header),
+                            });
+                        }
```

Comment:
Chatting with @Fraser999 , we should always do this

WDYT @fizyk20 ?

Reply:
What if someone misses to schedule an upgrade on time, and its node already managed to build blocks in an era where it should be running a new version? Node operator would try to fix its mistake by scheduling the upgrade again and IIRC, it would try to roll back those blocks. Is it possible to do without deleting them from the storage? In which case, such node would fail with the above error.

Reply:
In the circumstance you describe the code would never get here.  It would set `latest_block_header` in the `if let` branch at line 555 (see [here][1]).

[1]: https://github.com/casper-network/casper-node/pull/1796/files#diff-212250b2151062263c01ec4511602e2ada0b8a83dd02b8727589737f3d3312ebR555

Reply:
What should we always do? Bail if there exists a block higher than the current activation point? I'm not sure. I imagine we might _want_ to overwrite existing blocks sometimes, like when we're doing an emergency upgrade... though we'd probably use a hard reset in such cases and the offending blocks should be removed earlier, right?
Also, when would we _not_ have a `latest_block_header` if there are blocks in storage already?

Reply:
> Also, when would we not have a `latest_block_header` if there are blocks in storage already?

The joiner reactor might not have the latest block header if it was skipped (ie, in the case of an upgrade or genesis).